src/Lime.Transport.SignalR/ClientSignalRTransport.cs

```diff
@@ -0,0 +1,179 @@
+                    TraceWriter.TraceIfEnabledAsync(ex.ToString(), DataOperation.Error);
+                    return string.Empty;
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public override IReadOnlyDictionary<string, object> Options => new Dictionary<string, object>
+        {
+            [nameof(_hubConnection.HandshakeTimeout)] = _hubConnection.HandshakeTimeout,
+            [nameof(_hubConnection.ServerTimeout)] = _hubConnection.ServerTimeout,
+            [nameof(_hubConnection.KeepAliveInterval)] = _hubConnection.KeepAliveInterval
+        };
+
+        /// <inheritdoc />
+        /// <exception cref="ObjectDisposedException">This instance has already been disposed.</exception>
+        public override async Task<Envelope> ReceiveAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfNotConnected();
+
+            return await base.ReceiveAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        /// <exception cref="ObjectDisposedException">This instance has already been disposed.</exception>
+        public override async Task SendAsync(Envelope envelope, CancellationToken cancellationToken)
+        {
+            ThrowIfNotConnected();
+
+            await base.SendAsync(envelope, cancellationToken).ConfigureAwait(false);
+
+            string envelopeSerialized = EnvelopeSerializer.Serialize(envelope);
+            await TraceWriter.TraceIfEnabledAsync(envelopeSerialized, DataOperation.Send).ConfigureAwait(false);
+
+            await _hubConnection.SendAsync("FromClient", envelopeSerialized, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        /// <exception cref="ObjectDisposedException">This instance has already been disposed.</exception>
+        protected override async Task PerformCloseAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            await _hubConnection.StopAsync(cancellationToken).ConfigureAwait(false);
+            await base.PerformCloseAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        /// <exception cref="ObjectDisposedException">This instance has already been disposed.</exception>
+        protected override async Task PerformOpenAsync(Uri uri, CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            if (uri is null)
+            {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            await base.PerformOpenAsync(uri, cancellationToken).ConfigureAwait(false);
+
+            _hubConnection = new HubConnectionBuilder()
+                .WithUrl(uri.ToString() + "envelope")
+                .WithAutomaticReconnect()
+                .Build();
+
+            _hubConnection.On<string>("FromServer", async envelope =>
+            {
+                await EnvelopeChannel.Writer.WriteAsync(envelope);
+            });
```

Comment:
Return the underlying task to reduce the overhead. Also define a constant for the `"FromServer"` value.
```suggestion
            _hubConnection.On<string>(
               FROM_SERVER_METHOD, 
               envelope => EnvelopeChannel.Writer.WriteAsync(envelope).AsTask());
```

Reply:
@andrebires Do you know if this could have any unintended consequences here? I don't actually see how this would improve the code really. A single client won't send that much requests in a short time-frame and most of the time will be probably spent in network travel time, thus making the overhead that more insignificant. Even more so when considering that this is code that will be running in a single client -- as opposed to server-side `ReceiveAsync`, for example, which will be called hundreds of times a second by different clients in a single server instance -- and especially since this is the C# client, which is mostly used in testing scenarios. As it's [good practice](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task) to leave `async` anyway, I feel like this could be a case of premature optimization. Thoughts?